cli/file_watcher.rs

```diff
@@ -74,180 +77,127 @@ async fn error_handler(watch_future: FileWatcherFuture<Result<(), AnyError>>) {
   }
 }
 
-/// This function adds watcher functionality to subcommands like `fmt` or `lint`.
-/// The difference from [`watch_func_with_module_resolution`] is that this doesn't depend on
-/// [`ModuleGraph`].
-///
-/// - `target_resolver` is used for resolving file paths to be watched at every restarting of the watcher. The
-/// return value of this closure will then be passed to `operation` as an argument.
-///
-/// - `operation` is the actual operation we want to run every time the watcher detects file
-/// changes. For example, in the case where we would like to apply `fmt`, then `operation` would
-/// have the logic for it like calling `format_source_files`.
-///
-/// - `job_name` is just used for printing watcher status to terminal.
-///
-/// Note that the watcher will stop working if `target_resolver` fails at some point.
-///
-/// [`ModuleGraph`]: crate::module_graph::Graph
-pub async fn watch_func<F, G>(
-  target_resolver: F,
-  operation: G,
-  job_name: &str,
-) -> Result<(), AnyError>
+pub enum ResolutionResult<T> {
+  Restart {
+    paths_to_watch: Vec<PathBuf>,
+    result: Result<T, AnyError>,
+  },
+  Ignore,
+}
+
+async fn next_restart<F, T: Clone>(
+  resolver: &mut F,
+  debounce: &mut Pin<&mut Debounce>,
+  initial: bool,
+) -> (Vec<PathBuf>, Result<T, AnyError>)
 where
-  F: Fn() -> Result<Vec<PathBuf>, AnyError>,
-  G: Fn(Vec<PathBuf>) -> FileWatcherFuture<Result<(), AnyError>>,
+  F: FnMut(Option<Vec<PathBuf>>) -> FileWatcherFuture<ResolutionResult<T>>,
```

Comment:
If there's a `where` clause, I think it would be good to put all trait bounds there - I mean, I'd like to suggest that `T: Clone` be in `where` clause:

```rust
where
  F: FnMut(Option<Vec<PathBuf>>) -> FileWatcherFuture<ResolutionResult<T>>,
  T: Clone,
```

Reply:
Turns out `T` didn't actually need to be `Clone` in the first place - not sure why that was there.